### PR TITLE
feat: roll flows for counters — Spyrer Power Boost (#706 Phases 2+3)

### LINKED ISSUE #1427: Power Boost flow: roll table assignments + mod integration (Phase 2)
## Context

Design doc: https://github.com/gyrinx-app/gyrinx/discussions/1425
Parent issue: #706
Depends on: #1426

## Phase 2: Power Boost Flow

This phase adds the roll flow wizard, modifier integration, and cost propagation.

### New core model

- `ListFighterRollTableRowAssignment` (AppBase) -- links ListFighter to ContentRollTableRow. Fields: fighter FK, row FK, date_received, rating_increase (copied from row), notes, campaign_action (OneToOne to CampaignAction). This is the new 6th mod source.

### Modifier integration

- Add roll table results to `ListFighter._mods()` as the 6th mod source, following the injury pattern.
- In campaign mode, iterate `roll_table_results.all()` and extend with `row.modifiers.all()`.
- Stat changes appear automatically in the fighter statline via existing `_apply_mods()`.

### Cost propagation

- When assignment created: `propagate_from_fighter(fighter, Delta(delta=rating_increase, list=lst))`.
- When removed: propagate negative delta to reverse.
- Create ListAction recording the change.

### Roll flow UI

Entry point: button on the counter edit page (shown when counter value >= flow cost).

Views:
1. `roll_flow_start` -- show available flows for this counter
2. `roll_flow_roll` -- dice roll interface, show result from table
3. `roll_flow_confirm` -- review result details, apply

Handler: `handle_roll_flow()` -- transactional function that creates assignment, deducts counter, propagates cost, records CampaignAction.

### Power Boost removal

- Deletion reverses rating_increase and removes stat modifiers.
- Follows advancement deletion pattern with negative cost propagation.
- Creates ListAction with negative delta.

### Display

- Show Power Boost results on fighter card (similar to injuries/advancements section).

### Prefetch

- Add `roll_table_results`, `roll_table_results__row__modifiers` to `with_related_data()`.

### LINKED ISSUE #1428: Counters & roll tables: polish, performance, docs, tests (Phase 3)
## Context

Design doc: https://github.com/gyrinx-app/gyrinx/discussions/1425
Parent issue: #706
Depends on: #1426, #1427

## Phase 3: Polish

Final phase covering performance, documentation, and comprehensive testing.

### Prefetch optimisation

- Audit all prefetch paths for new relations (counters, roll_table_results).
- Ensure no N+1 queries on fighter card rendering with counters.
- Update performance query snapshots.

### Tests

- Content model tests (ContentCounter, ContentRollTable, ContentRollTableRow, ContentRollFlow).
- ListFighterCounter CRUD tests.
- ListFighterRollTableRowAssignment creation and deletion tests.
- Modifier integration tests (roll table results appear in _mods(), affect statline).
- Cost propagation tests (rating_increase applied and reversed correctly).
- Roll flow view tests (start, roll, confirm).
- Handler tests (handle_roll_flow transactional behaviour).
- Counter display tests (detection via content_fighter match).
- Counter edit tests (ListFighterCounter created on first save).

### Documentation

- Update docs/content-library/ with new models documentation.
- Follow existing content-library doc style.

### Edge cases

- Fighter with multiple counters.
- Counter value exactly at flow cost threshold.
- Multiple Power Boost results on same fighter (stacking modifiers).
- Fighter cloning with counters and roll table results.
- Archive/unarchive behaviour for counters and assignments.

### PR DESCRIPTION
**Pull request by tgvashworth**
Spyrers track Kill Count and Glitch Count, and can spend 4 Kill Count to roll on the Power Boost table for a permanent upgrade. Phase 1 (#1426) shipped the content models and counter display; this PR ships the remaining behaviour — the "spend a counter, roll on a table, gain the result" flow — plus the polish phase.

## Summary

- **Roll flow wizard**: from a counter's edit page, a fighter can start any flow defined for that counter (e.g. "Suit Evolution: spend 4 Kill Count, roll D6 on Power Boost"). The roll step supports an automatic roll or manual tabletop dice entry, records the dice as a `CampaignAction`, and the confirm step shows the matched table row (stat changes, cost increase) before applying it.
- **Results are permanent but removable**: a new `ListFighterRollResult` records the gained row, the counter points spent, and the rating increase (copied at gain time, so later content edits can't corrupt the books). Removing a result reverses the stat mods, reduces the fighter's cost, and refunds the counter — mirroring advancement deletion.
- **Counter warning highlight**: new `ContentCounter.warning_stat` turns the counter badge red when the value exceeds a statline stat — so Glitch Count goes red when it passes Toughness, per the "rig shuts down" rule.

- **Docs**: new content-library page [`docs/content-library/counters-and-roll-tables.md`](../blob/worktree-issue-706-roll-flow/docs/content-library/counters-and-roll-tables.md) with a Mermaid ER diagram of the models, field tables, app behaviour, and the Spyrer setup runbook (linked from the section README).
- **Warning badge affordance**: the warn state shows a dot after the value and uses the site's Bootstrap tooltip pattern (help cursor included) instead of a native `title`.

## Design notes

Re-derived against current `main` rather than the original Feb plan on #1427; the notable deviations:

- **Advancement pattern, not injury pattern.** Power Boosts are permanent purchased improvements, so their mods are *not* gated on campaign mode, and their cost participates in all three cost paths — `cost_int()`, `facts_from_db()`, and the push-propagation + `ListAction` ledger — so post-#1826 reconcile finds no drift. No pins (those are equipment-only).
- **Model named `ListFighterRollResult`** (design doc had `ListFighterRollTableRowAssignment`) — shorter, matches `ListFighterInjury`/`ListFighterAdvancement` naming.
- **D66 handled correctly**: `roll_value_from_dice` reads D66 as tens+units (11–66), not a sum. `roll_value` ranges like "2-3" are parsed tolerantly; malformed rows log and never match.
- **Clone carries roll results and counter values** — advancements already clone, and skipping results would make a clone's rating drift from its computed cost (the #1953 bug class).
- **Flow state is URL-driven** (`campaign_action_id` or `dice=` query param), like the advancement wizard — no session state, shareable/refreshable confirm step.
- **No content seeded in this repo.** Prod content is admin-managed (the 17 "Spyrer Glitches" injuries were entered that way); the runbook below covers setup.

## Content setup runbook (prod, via admin)

Nothing is seeded in prod (verified via prodshell — zero `ContentCounter`/`ContentRollTable`/`ContentRollFlow` rows). To activate for Spyrers:

1. **Counters** (Content → Counters):
   - "Kill Count", display order 0, restricted to the 7 Spyre Hunting Party fighters (Hunt Master, Jakara, Malcadon, Orrus, Sthenian, Yeld, Caryatid Prime — not the stash).
   - "Glitch Count", display order 1, same fighters, **warning stat: `T`**.
2. **Roll table** (Content → Roll Tables): "Power Boost", dice D6, one row per rulebook entry with `roll_value` ("1", "2" … or ranges like "1-2"), stat modifiers, and the rating increase from the book.
3. **Flow** (inline on the Kill Count counter, or Content → Roll Flows): "Suit Evolution" — counter: Kill Count, cost 4, table: Power Boost.

Glitches themselves remain the existing "Spyrer Glitches" injury table; users bump Glitch Count manually (auto-increment from injuries stays out of scope for v1, per the design doc).

## Test matrix

Full matrix in `gyrinx/core/tests/test_roll_flow.py` (34 tests):

- [x] Roll semantics: `parse_roll_value` (single/range/whitespace/reversed/malformed), D6/2D6/D66 value combination, `row_for_roll` matching incl. malformed-row tolerance
- [x] Handler: success path (deduction, copied fields, CampaignAction outcome, ListAction deltas), insufficient counter, stash fighter rejection, idempotent double-submit, no-campaign lists
- [x] Cost invariant: `cost_int()`, `rating_current` (push) and `facts_from_db()` (pull) all agree after gain and after removal
- [x] Deletion: archive + refund + rating reversal + negative ListAction; double-delete and wrong-fighter rejected
- [x] Mods: apply to statline (all list modes), stack, stop applying when archived
- [x] Views: permissions (owner/arbitrator/stranger/inapplicable flow), auto + manual roll, unaffordable block, dice-param fallback, confirm good/bad state, double-submit idempotency, results edit + remove
- [x] Card display: results grouped by table name; warning badge edge cases (no stat, unknown stat, equal-to-stat)
- [x] Clone: results + counters carried, costs stay in sync
- [x] Performance snapshot updated (70 → 72 queries; two new prefetch chains)
- [x] Full suite: 3110 passed / 13 skipped / 2 xfailed; bandit clean
- [x] Browser-verified end-to-end on a Spyre Hunting Party gang: roll → confirm → apply (rating 200¢→215¢, list 590¢→605¢, Kill Count 5→1, S 3→4, campaign log entries) → remove (everything reverts, counter refunded)

## How to try it locally

1. Seed content: create the two counters, a D6 table with rows + stat mods, and a flow (admin or shell) restricted to a fighter type in one of your lists.
2. Give a fighter of that type some counter value (fighter card → counter → Edit).
3. On the counter edit page, hit **Start** under "Spend Kill Count" → roll → apply, and watch the fighter card, rating, and campaign log update.

Closes #1427
Closes #1428
Refs #706

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: gyrinx/content/models/roll_table.py [hunk 1]
```diff
@@ -7,11 +7,42 @@
...
+    try:
+        if len(parts) == 1:
+            value = int(parts[0])
+            return (value, value)
+        if len(parts) == 2:
+            low, high = int(parts[0]), int(parts[1])
+            if low > high:
+                low, high = high, low
+            return (low, high)
+    except ValueError:
```
[the remark this answers is not in the document]
Comment (tgvashworth) [reply gyrinx/content/models/roll_table.py:42]: Fixed in 8c5bcec1.

File: gyrinx/core/views/fighter/roll_flow.py [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+                    with transaction.atomic():
+                        campaign_action = CampaignAction.objects.create(
+                            user=request.user,
+                            owner=request.user,
+                            campaign=lst.campaign,
+                            list=lst,
+                            description=description,
+                            dice_count=len(dice),
+                            dice_results=dice,
+                            dice_total=sum(dice),
```
[the remark this answers is not in the document]
Comment (tgvashworth) [reply gyrinx/core/views/fighter/roll_flow.py]: Fixed in c839b42f.

File: gyrinx/core/models/list/fighter.py [hunk 3]
```diff
@@ -2457,6 +2565,25 @@ def clone(self, **kwargs):
             advancement.campaign_action = None  # Clear campaign action reference
             advancement.save()
 
+        # Clone roll results — like advancements they carry a rating increase,
+        # so skipping them would make the clone's rating drift from its cost.
+        for roll_result in self.roll_results.all():
+            roll_result.pk = None
+            roll_result.fighter = clone
+            roll_result.campaign_action = None
+            roll_result.save()
```
[the remark this answers is not in the document]
Comment (tgvashworth) [reply gyrinx/core/models/list/fighter.py:2760]: Fixed in c839b42f.

File: gyrinx/core/handlers/fighter/roll_flow.py [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+            fighter_counter.save_with_user(user=user)
+        else:
+            # Unreachable unless flow.cost == 0, guarded above; kept for safety
+            ListFighterCounter.objects.create_with_user(
+                user=user,
+                fighter=fighter,
+                counter=flow.counter,
+                value=-flow.cost,
+                owner=lst.owner,
+            )
```
[the remark this answers is not in the document]
Comment (tgvashworth) [reply gyrinx/core/handlers/fighter/roll_flow.py]: Fixed in 248af778 — dead branch removed.

File: gyrinx/core/templates/core/list_fighter_roll_result_remove.html [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+          method="post">
+        {% csrf_token %}
+        <p>Are you sure you want to remove this result from {{ fighter.name }}?</p>
+        <div class="mt-3">
+            <button type="submit" class="btn btn-danger">Remove</button>
+            <a href="{% url 'core:list-fighter-roll-results-edit' list.id fighter.id %}"
+               class="btn btn-link">Cancel</a>
+        </div>
+    </form>
+</div>
```
[the remark this answers is not in the document]
Comment (tgvashworth) [reply gyrinx/core/templates/core/list_fighter_roll_result_remove.html:41]: False positive: the template's divs are balanced (6 opens / 6 closes, nesting depth never negative — verified programmatically after stripping template tags). djlint's indentation of the alert close is misleading but the DOM is correct; the page renders correctly in browser testing.

File: gyrinx/core/templates/core/list_fighter_roll_result_remove.html [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+            </div>
+        </div>
+        <div class="alert alert-info alert-icon mb-0" role="alert">
+            <i class="bi-info-circle"></i>
+            <div>
+                Removing this result will reverse its stat changes,
+                reduce {{ fighter.name }}'s cost by {{ roll_result.rating_increase }}¢{% if roll_result.counter_cost and roll_result.counter %},
+                and refund {{ roll_result.counter_cost }} {{ roll_result.counter.name }}{% endif %}.
+        </div>
+    </div>
```
[the remark this answers is not in the document]
Comment (tgvashworth) [reply gyrinx/core/templates/core/list_fighter_roll_result_remove.html:30]: False positive — see reply on the duplicate thread: tags are balanced; the indentation is djlint's canonical output.

File: gyrinx/core/templates/core/list_fighter_roll_result_remove.html [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+                <p class="text-secondary mb-0">Received: {{ roll_result.date_received|date:"M j, Y" }}</p>
+            </div>
+        </div>
+        <div class="alert alert-info alert-icon mb-0" role="alert">
+            <i class="bi-info-circle"></i>
+            <div>
+                Removing this result will reverse its stat changes,
+                reduce {{ fighter.name }}'s cost by {{ roll_result.rating_increase }}¢{% if roll_result.counter_cost and roll_result.counter %},
+                and refund {{ roll_result.counter_cost }} {{ roll_result.counter.name }}{% endif %}.
+        </div>
```
[the remark this answers is not in the document]
Comment (tgvashworth) [reply gyrinx/core/templates/core/list_fighter_roll_result_remove.html:29]: Deliberate: this mirrors the existing remove-injury and advancement templates, which use the same alert markup. Migrating all of them to the border-rounded pattern is a follow-up sweep, not this PR.

File: gyrinx/core/handlers/fighter/roll_flow.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+    # Link or create the CampaignAction
+    campaign_action = None
+    if campaign_action_id:
+        try:
+            campaign_action = CampaignAction.objects.get(id=campaign_action_id)
+        except CampaignAction.DoesNotExist:
+            raise ValidationError(f"Campaign action {campaign_action_id} not found")
+        campaign_action.outcome = outcome
+        campaign_action.save()
```
[the remark this answers is not in the document]
Comment (tgvashworth) [reply gyrinx/core/handlers/fighter/roll_flow.py:170]: Fixed in 1564f235 — lookup scoped to the fighter's list.

File: gyrinx/core/models/list/roll_result.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+import logging
+
+from django.db import models
+from simple_history.models import HistoricalRecords
+
+from gyrinx.core.models.base import AppBase
+from gyrinx.core.models.list.fighter import ListFighter
+
+logger = logging.getLogger(__name__)
+
```
[the remark this answers is not in the document]
Comment (tgvashworth) [reply gyrinx/core/models/list/roll_result.py]: Fixed in 1564f235.
